sklearn/model_selection/_search.py

```diff
@@ -94,7 +94,8 @@ class ParameterGrid:
     def __init__(self, param_grid):
         if not isinstance(param_grid, (Mapping, Iterable)):
             raise TypeError(
-                "Parameter grid is not a dict or a list ({!r})".format(param_grid)
+                f"Parameter grid should be a dict or a list, got: {param_grid!r} of"
+                f" type {type(param_grid)}"
```

Comment:
Maybe the following would lead to even nicer messages:

```suggestion
                f" type {type(param_grid).__name__}"
```

if you agree, some of the tests will need to be updated accordingly.